Add Support for Verifying Signature During Bundle Creation

ejoerns opened the pull request:
To detect signing errors and expired certificates early, this series adds trust chain verificiation during creation of a bundle if the  `--keyring` argument is provided or a keyring configured.

Additionally a check is implemented to warn the signer about using certificates in the chain that will expire within the period of a month.

jluebbe commented on src/signature.c:
```diff
@@ -180,6 +180,28 @@ GBytes *cms_sign(GBytes *content, const gchar *certfile, const gchar *keyfile, g
+	if (r_context()->config->keyring_path) {
+		CMS_ContentInfo *vcms = NULL;
+		X509_STORE *store = NULL;
+		STACK_OF(X509) *verified_chain = NULL;
```
Should `store` and `verified_chain` be freed?

ejoerns replied:
Ack. Used g_autoptr for these now.

jluebbe commented on docs/using.rst:
```diff
@@ -41,6 +41,13 @@ which image to install to which slot, which scripts to execute etc.
+While the ``--cert`` and ``--key`` argument are mandatory for signing and must
+provide the certificate and private key that should be used for creating the
+signature, the ``--keyring`` argument is optional and (if given) will be used
+for verifying the trust chain validity of the created signature after creation.
```
The 'created' before signature is redundant.

ejoerns replied:
removed

jluebbe commented on .travis.yml:
```diff
@@ -6,6 +6,7 @@ before_install:
   - sudo add-apt-repository "deb http://archive.ubuntu.com/ubuntu/ xenial main restricted universe multiverse" -y
   - sudo apt-get update -qq
   - sudo apt-get install -qq libglib2.0-dev libjson-glib-dev
+  - sudo apt-get install -qq faketime
```
If PR #300 is merged first, this should move to the normal packages list.

ejoerns replied:
That's the case, thus moved faketime to packages list.

jluebbe commented on src/signature.c:
```diff
@@ -353,6 +353,52 @@ GBytes *cms_sign(GBytes *content, const gchar *certfile, const gchar *keyfile, g
+	if (r_context()->config->keyring_path) {
+		g_autoptr(CMS_ContentInfo) vcms = NULL;
+		g_autoptr(X509_STORE) store = NULL;
+		STACK_OF(X509) *verified_chain = NULL;
```
Why is it not necessary to free verified_chain? A comment would help.

jluebbe commented on src/signature.c:
```diff
@@ -353,6 +353,52 @@ GBytes *cms_sign(GBytes *content, const gchar *certfile, const gchar *keyfile, g
+		}
+
+		for (int i = 0; i < sk_X509_num(verified_chain); i++) {
+			ASN1_TIME *expiry_time;
```
Do we need to free expiry_time?

ejoerns commented:
Pushed changes:
* Rebased onto master
* Resolved conflicts in rauc.t
* use X509_get0_not{Before,After} functions for clearer memory ownership, added backward compatibility to OpenSSL 1.0.x
* added missing free for `verified_chain`
* exemplarily documented memory ownership of `cms_get_cert_chain()`
